### PR TITLE
refactor(fingerprint): Drop trivial `p-limit` dependency

### PR DESCRIPTION
**Pull request by kitten**
# Why

This won't have the biggest impact, but in theory, it may be dropped when `@expo/xcpretty` is updated to drop `find-up` and Jest & Eslint aren't in use in a project. Mainly, the motivation here is to remove it, since it's trivially replaced.

# How

- Add `utils/Concurrency.ts` with `createLimiter` utility and tests
- Replace `p-limit` with `createLimiter`
- Fix some typing issues

# Test Plan

- Tests added for `createLimiter`
- Existing tests and E2E tests should pass

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

### CODE REVIEW

**Review by hassankhan — APPROVED** [review]

File: packages/@expo/fingerprint/src/hash/__tests__/Hash-ReactImportsPatcher-test.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import pLimit from 'p-limit';
 import path from 'path';
 
 import { normalizeOptionsAsync } from '../../Options';
+import { createLimiter } from '../../utils/Concurrency';
```
Comment (Kudo) [inline packages/@expo/fingerprint/src/hash/__tests__/Hash-ReactImportsPatcher-test.ts:5]: [🚨 **critical**] **critical**: Incomplete migration — `p-limit` has been removed from `package.json` but is still imported on line 3 (`import pLimit from 'p-limit'`). Two test cases below still use `pLimit(1)` instead of `createLimiter(1)` (lines 51 and 73: `"should not use ReactImportsPatchTransform if no match platforms"` and `"should not use ReactImportsPatchTransform if no match files"`). These tests will break once a hoisted copy of `p-limit` is no longer available in the monorepo.

The fix is straightforward — replace the remaining `pLimit(1)` calls with `createLimiter(1)` and remove the `p-limit` import.

File: packages/@expo/fingerprint/src/utils/Concurrency.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    ...args: Arguments
+  ): Promise<ReturnType>;
+}
+
+interface QueueItem {
+  resolve(): void;
+  next: QueueItem | null;
+}
+
+export const createLimiter = (limit = 1): Limiter => {
```
Comment (Kudo) [inline packages/@expo/fingerprint/src/utils/Concurrency.ts:13]: [💡 suggestion] **suggestion**: No input validation on `limit`. If someone passes `0`, `NaN`, or a negative number, the limiter will either deadlock (limit=0) or allow unbounded concurrency (negative). `p-limit` throws a `TypeError` for these cases.

Consider adding a guard, e.g.:
```ts
if (!(Number.isInteger(limit) && limit > 0) && limit !== Infinity) {
  throw new TypeError('Expected `limit` to be a positive integer or Infinity');
}
```

This is a minor concern since the only caller passes `options.concurrentIoLimit` which has a validated default, but it would make the utility more robust if reused elsewhere.

File: packages/@expo/fingerprint/src/utils/__tests__/Concurrency-test.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  return {
+    promise,
+    resolve(value: T): Promise<void> {
+      resolve(value);
+      return promise.then(() => Promise.resolve());
+    },
+  };
+};
+
+describe(createLimiter, () => {
```
Comment (Kudo) [inline packages/@expo/fingerprint/src/utils/__tests__/Concurrency-test.ts:17]: [💡 suggestion] **suggestion**: The test covers the core concurrency limiting behavior well. A few additional cases would strengthen coverage:
- Error propagation: verify that a rejected promise from `fn` is properly propagated and the slot is released for queued work.
- `limit=1` (mutex behavior): verify strict serial execution.
- Synchronous `fn`: verify that a non-async function works correctly.

These are all paths exercised by the production code (e.g., `createFileHashResultsAsync` passes a synchronous-returning callback to the limiter).

**Review by Kudo — CHANGES_REQUESTED** [review]
Replaces the `p-limit` dependency with a custom `createLimiter` utility in `@expo/fingerprint`. The implementation uses a hand-rolled singly-linked list queue and is correct for valid inputs. The synchronous fast-path admission control is actually cleaner than p-limit's design (which needs an `await Promise.resolve()` workaround). Zero new dependencies.

However, **two test cases in `Hash-ReactImportsPatcher-test.ts` were not migrated** — they still import and call `pLimit` from `p-limit`, which has been removed from `package.json`. This will cause test failures once hoisted copies of `p-limit` are unavailable.

Additionally, the `createLimiter` implementation lacks input validation for edge cases like `limit=0` or negative values, unlike `p-limit` which throws a `TypeError`.

**Review by Kudo — APPROVED** [review]
all nit comments from my review skill 😅
approve it and let you follow up/merge

File: packages/@expo/fingerprint/src/utils/Concurrency.ts (hunk 2 above)
Comment (kitten) [reply packages/@expo/fingerprint/src/utils/Concurrency.ts:13]: Protecting against `NaN` for an internal param is excessive and the failure would become immediately obvious

File: packages/@expo/fingerprint/src/utils/__tests__/Concurrency-test.ts (hunk 3 above)
Comment (kitten) [reply packages/@expo/fingerprint/src/utils/__tests__/Concurrency-test.ts:17]: > These are all paths exercised by the production code

I think that's right, yes, so the AI comment basically agrees that this is covered via other tests 🤔 Will have a look after merging

File: packages/@expo/fingerprint/src/hash/__tests__/Hash-ReactImportsPatcher-test.ts (hunk 1 above)
Comment (kitten) [reply packages/@expo/fingerprint/src/hash/__tests__/Hash-ReactImportsPatcher-test.ts:5]: very red 😆 yes, I missed this
